subsys/nrf_compress/src/lzma.c

```diff
@@ -303,6 +422,260 @@ static int lzma_decompress(void *inst, const uint8_t *input, size_t input_size,
+
+		allocated_probs = true;
+#ifdef CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2
+		*offset = LZMA2_HEADER_SIZE;
+#else
+		/* Header and account for uncompressed size */
+		*offset = LZMA_PROPS_SIZE + sizeof(uint64_t);
+#endif
+
+#ifdef CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2
+		Lzma2Dec_Init(&lzma_decoder);
+#else
+		LzmaDec_Init(&lzma_decoder);
+#endif
+
+		return 0;
+	}
+
+#ifdef CONFIG_NRF_COMPRESS_LZMA_VERSION_LZMA2
+	rc = Lzma2Dec_DecodeToDic(&lzma_decoder, MAX_LZMA_DICT_SIZE, input, &chunk_size,
+					(last_part ? LZMA_FINISH_END : LZMA_FINISH_ANY), &status);
+#else
+	rc = LzmaDec_DecodeToDic(&lzma_decoder, MAX_LZMA_DICT_SIZE, input, &chunk_size,
+					(last_part ? LZMA_FINISH_END : LZMA_FINISH_ANY), &status);
+#endif
+	if (rc) {
+		rc = -EINVAL;
+		goto done;
+	}
+
+	*offset = chunk_size;
+
+	if (last_part && (status == LZMA_STATUS_FINISHED_WITH_MARK ||
+			  status == LZMA_STATUS_MAYBE_FINISHED_WITHOUT_MARK) &&
+	    *offset < input_size) {
+		/* If last block, ensure offset matches complete file size */
+		*offset = input_size;
+	}
+
+	if (decoder->dicPos >= decoder->dicHandle->dicBufSize ||
+	    (last_part && input_size == chunk_size)) {
+#if CONFIG_NRF_COMPRESS_DICTIONARY_CACHE_SIZE > 0
+		if (cache.invalid) {
+			validate_cache(decoder->dicHandle);
+		}
+#endif
+		*output_size = decoder->dicPos;
+		decoder->dicPos = 0;
+	}
+
+done:
+	return rc;
+}
+
+static DictHandle dict_handle;
+
+DictHandle *LzmaDictionaryOpen(SizeT size)
+{
+	if (dict_handle.isOpened)
+		return &dict_handle;
+
+	if (ext_dict ==  NULL)
+		return NULL;
+
+	size_t dict_size;
+
+	if (ext_dict->open((size_t)size, &dict_size) != 0) {
+		LOG_WRN("Unable to open external dictionary with size %u", size);
+		return NULL;
+	}
+
+	dict_handle.isOpened = True;
+	dict_handle.dicBufSize = dict_size;
+
+#if CONFIG_NRF_COMPRESS_DICTIONARY_CACHE_SIZE > 0
+	cache.dict_pos_begin = 0;
+	cache.dict_pos_end = sizeof(cache.data) - 1;
+#endif
+
+	return &dict_handle;
+}
+
+SizeT LzmaDictionaryWrite(DictHandle *handle, SizeT pos, const Byte *data, SizeT len)
+{
+	if (handle != &dict_handle || ext_dict == NULL)
+		return 0;
+
+	SizeT write_len = len;
+
+	if (pos + len > handle->dicBufSize) {
+		write_len = handle->dicBufSize - pos;
+	}
+
+#if CONFIG_NRF_COMPRESS_DICTIONARY_CACHE_SIZE > 0
+	if (pos > cache.dict_pos_end || pos < cache.dict_pos_begin) {
+		/*
+		 * Should never happen, lzma operates on dicPos when writing to dictionary,
+		 * which should be aligned with cache.
+		 */
+		return 0;
+	}
+
+	SizeT bytes_written = 0;
+	SizeT cache_pos = pos - cache.dict_pos_begin;
+
+	while (bytes_written != write_len) {
+		SizeT cache_write_len =
+			(write_len - bytes_written) > (sizeof(cache.data) - cache_pos) ?
+				(sizeof(cache.data) - cache_pos) :
+				(write_len - bytes_written);
+
+		memcpy(cache.data + cache_pos, data + bytes_written, cache_write_len);
+		cache.invalid = true;
+
+		bytes_written += cache_write_len;
+		cache_pos += cache_write_len;
+
+		if (cache_pos == sizeof(cache.data)) {
```

Comment:
```suggestion
		if (cache_pos >= sizeof(cache.data)) {
```

Seems a bit safer, even though probably impossible that `cache_pos` is greater